src/v/cluster/rm_partition_frontend.cc

```diff
@@ -60,28 +62,127 @@ ss::future<begin_tx_reply> rm_partition_frontend::begin_tx(
+        if (!leader_opt) {
+            error = fmt::format("can't find {} in the leaders cache", ntp);
+            vlog(
+              clusterlog.trace,
+              "can't find {} in the leaders cache, retries left: {}",
+              ntp,
+              retries);
+            aborted = !co_await sleep_abortable(delay_ms);
+            leader_opt = _leaders.local().get_leader(ntp);
+            continue;
+        }
+
+        auto leader = leader_opt.value();
+        auto _self = _controller->self();
+
+        begin_tx_reply result;
+        if (leader == _self) {
+            result = co_await do_begin_tx(
+              ntp, pid, tx_seq, transaction_timeout_ms);
+            if (result.ec == tx_errc::leader_not_found) {
+                error = fmt::format(
+                  "local execution of begin_tx({},...) failed with 'not a "
+                  "leader'",
+                  ntp);
+                vlog(
+                  clusterlog.trace,
+                  "local execution of begin_tx({},...) failed with 'not a "
+                  "leader', retries left: {}",
+                  ntp,
+                  retries);
+                aborted = !co_await sleep_abortable(delay_ms);
+                leader_opt = _leaders.local().get_leader(ntp);
+                continue;
+            }
+            if (result.ec != tx_errc::none) {
+                vlog(
+                  clusterlog.warn,
+                  "local execution of begin_tx({},...) failed with {}",
+                  ntp,
+                  result.ec);
+            }
+            co_return result;
+        }
+
+        vlog(
+          clusterlog.trace,
+          "dispatching begin_tx({},...) from {} to {}",
+          ntp,
+          _self,
+          leader);
+        result = co_await dispatch_begin_tx(
+          leader, ntp, pid, tx_seq, transaction_timeout_ms, timeout);
+        if (result.ec == tx_errc::leader_not_found) {
+            error = fmt::format(
+              "remote execution of begin_tx({},...) on {} failed with 'not a "
+              "leader'",
+              ntp,
+              leader);
+            vlog(
+              clusterlog.trace,
+              "remote execution of begin_tx({},...) on {} failed with 'not a "
+              "leader', retries left: {}",
+              ntp,
+              leader,
+              retries);
+            aborted = !co_await sleep_abortable(delay_ms);
+            leader_opt = _leaders.local().get_leader(ntp);
+            continue;
+        }
+        if (result.ec != tx_errc::none) {
+            vlog(
+              clusterlog.warn,
+              "remote execution of begin_tx({},...) on {} failed with {}",
+              ntp,
+              leader,
+              result.ec);
+        }
+        co_return result;
+    }
+    if (error) {
+        vlog(clusterlog.warn, "{}", error.value());
+    }
+    co_return begin_tx_reply{.ntp = ntp, .ec = tx_errc::leader_not_found};
```

Comment:
is this always leader not found here when we get to the end?

Reply:
right